### PR TITLE
Simplify trigger functions in cacher

### PR DESCRIPTION
**Pull request by wojtek-t**
This is also work towards (potentially) supporting some indexes for lists in cacher.

Ref https://github.com/kubernetes/kubernetes/issues/73958

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go [hunk 1]
```diff
@@ -53,16 +53,16 @@ func StorageWithCacher(capacity int) generic.StorageDecorator {
...
-			Codec:                storageConfig.Codec,
+			CacheCapacity:         capacity,
+			Storage:               s,
+			Versioner:             etcd3.APIObjectVersioner{},
+			ResourcePrefix:        resourcePrefix,
+			KeyFunc:               keyFunc,
+			NewFunc:               newFunc,
+			NewListFunc:           newListFunc,
+			GetAttrsFunc:          getAttrsFunc,
+			TriggerPublisherFuncs: triggerFuncs,
```
Comment (yastij) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go]: do we plan to support indexing based on arbitrary fields ?

File: pkg/registry/core/pod/storage/storage.go [hunk 2]
```diff
@@ -78,7 +78,11 @@ func NewStorage(optsGetter generic.RESTOptionsGetter, k client.ConnectionInfoGet
 
 		TableConvertor: printerstorage.TableConvertor{TableGenerator: printers.NewTableGenerator().With(printersinternal.AddHandlers)},
 	}
-	options := &generic.StoreOptions{RESTOptions: optsGetter, AttrFunc: pod.GetAttrs, TriggerFunc: pod.NodeNameTriggerFunc}
+	options := &generic.StoreOptions{
+		RESTOptions: optsGetter,
+		AttrFunc:    pod.GetAttrs,
+		TriggerFunc: map[string]storage.TriggerPublisherFunc{"spec.nodeName": pod.NodeNameTriggerFunc},
```
Comment (yastij) [inline pkg/registry/core/pod/storage/storage.go]: this PR would enable indexing by `spec.nodeName`, will this indexing increase on the apiserver ?

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go]: no (at least not anytime in the near/mid future)

File: pkg/registry/core/pod/storage/storage.go (hunk 2 above)
Comment (wojtek-t) [reply pkg/registry/core/pod/storage/storage.go]: this PR is noop from functionality POV - we were already doing that

Comment (yastij) [reply pkg/registry/core/pod/storage/storage.go]: ack

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 3]
```diff
@@ -73,16 +73,13 @@ type ResponseMeta struct {
...
 }
 
-// MatchValue defines a pair (<index name>, <value for that index>).
-type MatchValue struct {
-	IndexName string
-	Value     string
-}
+// TriggerPublisherFunc is a function that for a given object computes
+// <value of an index> for a particular <index>.
+type TriggerPublisherFunc func(obj runtime.Object) string
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go]: Can we rename this to something like `IndexerFunc`? I didn't find the current name helpful in understanding what the function does (same for `TriggerPublisherFuncs` below).

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -298,15 +303,34 @@ func NewCacherFromConfig(config Config) *Cacher {
 		panic("storage codec doesn't seem to match given type: " + err.Error())
 	}
 
+	var indexedTrigger *indexedTriggerFunc
+	if config.TriggerPublisherFuncs != nil {
+		// For now, we don't support multiple trigger functions defined
+		// for a given resource.
+		// So until that is fixed, we choose only a random one and log error
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: How are triggers configured? I'd prefer we somehow reject attempts to configure the system more than one per resource. Randomly picking one and logging an error which might go unnoticed worries me.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: I opened https://github.com/kubernetes/kubernetes/pull/80160 to make that simpler - will fix once that one merges.

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go]: Added a TODO - will fix in a subsequent PR.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: Done

**Review by jpbetz — APPROVED** [review]
Looks to correctly introduce a map of trigger funcs without changing existing functionality.

/lgtm

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 5]
```diff
@@ -73,16 +73,15 @@ type ResponseMeta struct {
...
-	Value     string
-}
-
-// TriggerPublisherFunc is a function that takes an object, and returns a list of pairs
-// (<index name>, <index value for the given object>) for all indexes known
-// to that function.
-type TriggerPublisherFunc func(obj runtime.Object) []MatchValue
+// TriggerPublisherFunc is a function that for a given object computes
+// <value of an index> for a particular <index>.
+// TODO(wojtek-t): Rename to IndexerFunc?
```
Comment (tedyu) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go]: I can do this renaming if @wojtek-t hasn't started yet.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go]: If you have capacity, please do.

Comment (tedyu) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go]: See #80300

**Pull request merged**